Std/Data/String/Basic.lean

```diff
@@ -9,29 +9,52 @@ import Std.Data.Nat.Lemmas
 
 namespace Substring
 
+/-- Check if two substrings are equal as strings. -/
+def eqAsString (s t : Substring) : Bool :=
+  if s.bsize ≠ t.bsize then
+    false
+  else
+    go s.bsize 0
+where
+  /-- Auxiliary definition for `eqAsString`. -/
+  go (remaining : Nat) (offset : String.Pos) : Bool :=
+    if h : remaining = 0 then
+      true
+    else
+      let cs := s.get (s.startPos + offset)
+      let ct := t.get (t.startPos + offset)
+      if cs == ct then
+        have : remaining - String.csize cs < remaining :=
+          Nat.sub_lt (Nat.pos_of_ne_zero h) (String.csize_pos _)
+        go (remaining - String.csize cs) (offset + cs)
+      else
+        false
+
 /--
 If `pre` is a prefix of `s`, i.e. `s = pre ++ t`, return the remainder `t`.
 -/
 def dropPrefix? (s : Substring) (pre : Substring) : Option Substring :=
-  go 0 (pre.stopPos - pre.startPos)
-where
-  /-- Auxiliary definition for `dropPrefix?`. -/
-  go (start : String.Pos) (stop : String.Pos) : Option Substring :=
-    if h : start ≥ stop then
-      some { s with startPos := s.startPos + start }
+  if s.bsize < pre.bsize then
+    none
+  else
+    let p := s.startPos + (pre.stopPos - pre.startPos)
+    if { s with stopPos := p }.eqAsString pre then
+      some { s with startPos := p }
     else
-      let cs := s.get start
-      let cp := pre.get start
-      if cs == cp then
-        have : start.byteIdx + String.csize cs ≤ stop.byteIdx + 4 :=
-          Nat.add_le_add (Nat.le_of_lt $ Nat.not_le.mp h) (String.csize_le_4 _)
-        have : stop.byteIdx + 4 - (start.byteIdx + String.csize cs) <
-                stop.byteIdx + 4 - start.byteIdx :=
-          Nat.sub_add_lt_sub this (String.csize_pos _)
-        go (start + cs) stop
-      else
-        none
-termination_by go start stop => stop.byteIdx + 4 - start.byteIdx
+      none
+
+/--
+If `suff` is a suffix of `s`, i.e. `s = t ++ suff`, return the remainder `t`.
+-/
+def dropSuffix? (s : Substring) (suff : Substring) : Option Substring :=
+  if s.bsize < suff.bsize then
+    none
+  else
+    let p := s.stopPos + (suff.stopPos - suff.startPos)
```

Comment:
The implementation of `dropPrefix?` is wrong because it uses `+` here instead of `-`:
```lean
#eval "ab".toSubstring.dropSuffix? "b".toSubstring -- none
```
But even after fixing that there is a more subtle issue which makes it still wrong:
```lean
def dropSuffix? (s : Substring) (suff : Substring) : Option Substring :=
  if s.bsize < suff.bsize then
    none
  else
    let p := s.stopPos - (suff.stopPos - suff.startPos)
    if { s with startPos := p }.eqAsString suff then
      some { s with stopPos := p }
    else
      none

#eval "←".toSubstring.dropSuffix? "A".toSubstring -- some "←".toSubstring
```
The issue is that `p` is not necessarily a UTF8 code point boundary of `s`, and this is a *precondition* of `String.substrEq` / `eqAsString`. Instead you have to walk backward from the end of the string, as the implementation of `s.endsWith suff` does.

Reply:
May I recommend using these two functions to do the necessary comparisons:

```lean
protected theorem String.ne_zero_of_lt : {a b : Pos} → a < b → b ≠ 0
| _, _, hlt, rfl => Nat.not_lt_zero _ hlt

/-- Return the longest common prefix of two substrings. -/
def Substring.commonPrefix (s t : Substring) : Substring :=
  { s with stopPos := loop s.startPos t.startPos }
where
  loop spos tpos :=
    if h : spos < s.stopPos ∧ tpos < t.stopPos then
      if s.get spos == t.get tpos then
        have := Nat.sub_lt_sub_left h.1 (s.str.lt_next spos)
        loop (s.str.next spos) (t.str.next tpos)
      else
        spos
    else
      spos
termination_by loop => s.stopPos.byteIdx - spos.byteIdx

/-- Return the longest common suffix of two substrings. -/
def Substring.commonSuffix (s t : Substring) : Substring :=
  { s with startPos := loop s.stopPos t.stopPos }
where
  loop spos tpos :=
    if h : s.startPos < spos ∧ t.startPos < tpos then
      let spos' := s.str.prev spos
      let tpos' := t.str.prev tpos
      if s.get spos' == t.get tpos' then
        have : spos' < spos := s.str.prev_lt_of_pos spos (String.ne_zero_of_lt h.1)
        loop spos' tpos'
      else
        spos
    else
      spos
termination_by loop => spos.byteIdx
```

Reply:
Actually, these are still incorrect. Comparing `bsize` is not sufficient because of [overlong encodings](https://en.wikipedia.org/wiki/UTF-8#Overlong_encodings).

Reply:
Here is a correct version:
```lean
protected theorem String.ne_zero_of_lt : {a b : Pos} → a < b → b ≠ 0
| _, _, hlt, rfl => Nat.not_lt_zero _ hlt

private def Substring.commonPrefixAux (s t : Substring) : String.Pos × String.Pos :=
  loop s.startPos t.startPos
where
  loop spos tpos :=
    if h : spos < s.stopPos ∧ tpos < t.stopPos then
      if s.str.get spos == t.str.get tpos then
        have := Nat.sub_lt_sub_left h.1 (s.str.lt_next spos)
        loop (s.str.next spos) (t.str.next tpos)
      else
        (spos - s.startPos, tpos - t.startPos)
    else
      (spos - s.startPos, tpos - t.startPos)
termination_by loop => s.stopPos.byteIdx - spos.byteIdx

private def Substring.commonSuffixAux (s t : Substring) : String.Pos × String.Pos :=
  loop s.stopPos t.stopPos
where
  loop spos tpos :=
    if h : s.startPos < spos ∧ t.startPos < tpos then
      let spos' := s.str.prev spos
      let tpos' := t.str.prev tpos
      if s.str.get spos' == t.str.get tpos' then
        have : spos' < spos := s.str.prev_lt_of_pos spos (String.ne_zero_of_lt h.1)
        loop spos' tpos'
      else
        (spos - s.startPos, tpos - t.startPos)
    else
      (spos - s.startPos, tpos - t.startPos)
termination_by loop => spos.byteIdx

/-- Return the longest common prefix of two substrings. -/
def Substring.commonPrefix (s t : Substring) : Substring :=
  { s with stopPos := (commonPrefixAux s t).1 + s.startPos }

/-- Return the longest common suffix of two substrings. -/
def Substring.commonSuffix (s t : Substring) : Substring :=
  { s with startPos := (commonSuffixAux s t).1 + s.startPos }

/--
If `pre` is a prefix of `s`, i.e. `s = pre ++ t`, return the remainder `t`.
-/
def Substring.dropPrefix? (s : Substring) (pre : Substring) : Option Substring :=
  let (spos, tpos) := commonPrefixAux s pre
  if tpos.byteIdx = pre.bsize then
    some { s with startPos := spos + s.startPos }
  else
    none

/--
If `suff` is a suffix of `s`, i.e. `s = t ++ suff`, return the remainder `t`.
-/
def Substring.dropSuffix? (s : Substring) (suff : Substring) : Option Substring :=
  let (spos, tpos) := commonSuffixAux s suff
  if tpos = 0 then
    some { s with stopPos := spos + s.startPos }
  else
    none
```

Reply:
> Actually, these are still incorrect. Comparing `bsize` is not sufficient because of [overlong encodings](https://en.wikipedia.org/wiki/UTF-8#Overlong_encodings).

Overlong encodings are not encodings (the name is misleading). Those bit patterns are not legal UTF-8, and you are guaranteed not to see them in `String`s. A `a` character is always encoded in UTF-8 in one byte.

Reply:
Okay, switched to @fgdorais' implementation, thank you!